File: SimpleTunnelSSHForLinux/src/SimpleTunnelSSH.Core/Services/PortConflictDetector.cpp
#include "Services/PortConflictDetector.h"

namespace SimpleTunnelSSH::Core::Services
{

QList<Models::PortConflict> PortConflictDetector::detect(const QList<Models::TunnelGroup>& groups)
{
    QList<Models::PortConflict> conflicts;

    for (int firstIndex = 0; firstIndex < groups.size(); ++firstIndex)
    {
        const auto& firstGroup = groups[firstIndex];

        if (!firstGroup.isEnabled)
        {
            continue;
        }

        for (int secondIndex = firstIndex + 1; secondIndex < groups.size(); ++secondIndex)
        {
            const auto& secondGroup = groups[secondIndex];

            if (!secondGroup.isEnabled)
            {
                continue;
            }

            for (const auto& firstPort : firstGroup.ports)
            {
                if (!firstPort.isEnabled)
                {
                    continue;
                }

                for (const auto& secondPort : secondGroup.ports)
                {
                    if (!secondPort.isEnabled || firstPort.localPort != secondPort.localPort)
                    {
                        continue;
                    }

                    Models::PortConflict conflict;
                    conflict.firstGroupId = firstGroup.id;
                    conflict.firstGroupName = firstGroup.name;
                    conflict.secondGroupId = secondGroup.id;
                    conflict.secondGroupName = secondGroup.name;
                    conflict.localPort = firstPort.localPort;
                    conflicts.append(conflict);
                }
            }
        }
    }

    return conflicts;
}

} // namespace SimpleTunnelSSH::Core::Services

File: SimpleTunnelSSHForWindows/src/SimpleTunnelSSH.Core/Services/PortConflictDetector.cs
using SimpleTunnelSSH.Core.Models;

namespace SimpleTunnelSSH.Core.Services;

public static class PortConflictDetector
{
    public static IReadOnlyList<PortConflict> Detect(IEnumerable<TunnelGroup> groups)
    {
        var conflicts = new List<PortConflict>();
        var seenPorts = new Dictionary<int, (Guid GroupId, string GroupName)>();

        foreach (var group in groups.Where(static currentGroup => currentGroup.IsEnabled))
        {
            foreach (var port in group.Ports.Where(static currentPort => currentPort.IsEnabled))
            {
                if (seenPorts.TryGetValue(port.LocalPort, out var previous))
                {
                    if (previous.GroupId != group.Id)
                    {
                        conflicts.Add(new PortConflict(
                            port.LocalPort,
                            previous.GroupId,
                            previous.GroupName,
                            group.Id,
                            group.Name));
                    }

                    continue;
                }

                seenPorts[port.LocalPort] = (group.Id, group.Name);
            }
        }

        return conflicts;
    }
}